lib/netext/httpext/request.go

```diff
@@ -162,9 +162,90 @@ func compressBody(algos []CompressionType, body io.ReadCloser) (io.Reader, int64
 	return buf, int64(buf.Len()), contentEncoding, body.Close()
 }
 
+func readResponseBody(
+	state *lib.State, respType ResponseType, resp *http.Response, respErr error,
+) (interface{}, error) {
+
+	if resp == nil || respErr != nil {
+		return nil, respErr
+	}
+
+	// Transperently decompress the body if it's has a content-encoding we
+	// support. If not, simply return it as it is.
+	contentEncoding := strings.TrimSpace(resp.Header.Get("Content-Encoding"))
+	if compression, err := CompressionTypeString(contentEncoding); err == nil {
+		switch compression {
+		case CompressionTypeDeflate:
+			resp.Body, respErr = zlib.NewReader(resp.Body)
+		case CompressionTypeGzip:
+			resp.Body, respErr = gzip.NewReader(resp.Body)
+		default:
+			// We have not implemented a compression ... :(
+			respErr = fmt.Errorf(
+				"unsupported compression type %s for decompression - this is a bug in k6, please report it",
+				compression,
+			)
+		}
+	}
+
+	if respType == ResponseTypeNone {
+		_, err := io.Copy(ioutil.Discard, resp.Body)
+		_ = resp.Body.Close()
+		if err != nil && err != io.EOF {
+			respErr = err
+		}
+		return nil, respErr
+	}
+
+	buf := state.BPool.Get()
+	buf.Reset()
+	_, err := io.Copy(buf, resp.Body)
+	_ = resp.Body.Close()
+	if err != nil && err != io.EOF {
+		respErr = err
+	}
+
+	var result interface{}
+	// Binary or string
+	switch respType {
+	case ResponseTypeText:
+		result = buf.String()
+		state.BPool.Put(buf)
+	case ResponseTypeBinary:
+		result = buf.Bytes()
+		// We don't return the buffer to the pool, because the byte
+		// slice could still point to its underlying slice.
```

Comment:
Just out of curiosity, does the performance worsen if you copy the bytes here and return the buffer ?

Reply:
hmm good question... I'd expect it to worsen slightly, just because with that extra unnecessary copy you should loose any benefits from actually having the buffers, but I'll test it and report back

Reply:
I think it will get better ... slightly, because now you get a buffer you raise it size to something and than you throw it away . Then for the next request you need to this whole raising of the bucket size again. If you just reuse the bucket but copy the byte slice you are just copying the byte slice once and reuse the buffer ;)